// @azure/attestation/dist-esm/src/generated/generatedClient.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { PolicyImpl, PolicyCertificatesImpl, AttestationImpl, SigningCertificatesImpl, MetadataConfigurationImpl } from "./operations";
import { GeneratedClientContext } from "./generatedClientContext";
export class GeneratedClient extends GeneratedClientContext {
    /**
     * Initializes a new instance of the GeneratedClient class.
     * @param instanceUrl The attestation instance base URI, for example https://mytenant.attest.azure.net.
     * @param options The parameter options
     */
    constructor(instanceUrl, options) {
        super(instanceUrl, options);
        this.policy = new PolicyImpl(this);
        this.policyCertificates = new PolicyCertificatesImpl(this);
        this.attestation = new AttestationImpl(this);
        this.signingCertificates = new SigningCertificatesImpl(this);
        this.metadataConfiguration = new MetadataConfigurationImpl(this);
    }
}
