// @azure/attestation/dist-esm/src/generated/operations/signingCertificates.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers";
import * as Parameters from "../models/parameters";
/** Class representing a SigningCertificates. */
export class SigningCertificatesImpl {
    /**
     * Initialize a new instance of the class SigningCertificates class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Retrieves metadata signing certificates in use by the attestation service
     * @param options The options parameters.
     */
    get(options) {
        return this.client.sendOperationRequest({ options }, getOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getOperationSpec = {
    path: "/certs",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.JsonWebKeySet
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    urlParameters: [Parameters.instanceUrl],
    headerParameters: [Parameters.accept2],
    serializer
};
